server/app/middlewares/websocket_backend.rb

```diff
@@ -213,17 +212,16 @@ def on_open(ws, req)
     EM.defer { Agent::NodePlugger.new(node).plugin! connected_at }
 
   rescue CloseError => exc
-    logger.warn "reject websocket connection: #{exc}"
+    logger.warn "reject websocket connection for node #{node || node_id || '<nil>'}: #{exc}"
     ws.close(exc.code, exc.message)
 
+    # XXX: this only applies to the version/clock skew errors, not the token -> node errors
+    Agent::NodePlugger.new(node).reject!(connected_at, exc.code, exc.message) if node
```

Comment:
This needs to use something like `min(connected_at, Time.now)` for the case of a agent sending a `connected_at` timestamp in the future, to avoid bumping the `HostNode.connected_at` to an incorrect timestamp. This would prevent the agent from re-connecting with the correct timestamp, until the actual wallclock time catches up with the invalid `connected_at` timestamp in the future...